src/cpp/session/modules/environment/SessionEnvironment.cpp

```diff
@@ -112,6 +112,57 @@ bool handleRBrowseEnv(const core::FilePath& filePath)
    }
 }
 
+bool hasExternalPtr(SEXP env,      // environment to search for external pointers
+                    bool nullPtr,  // whether to look for NULL pointers 
+                    int level = 5) // maximum recursion depth (envs can have self-ref loops)
+{
+   // list the contents of this environment
+   std::vector<r::sexp::Variable> vars;
+   r::sexp::Protect rProtect;
+   r::sexp::listEnvironment(env, 
+                            true,  // include all values
+                            false, // don't include last dot
+                            &rProtect, &vars);
+
+   // check for external pointers
+   for (std::vector<r::sexp::Variable>::iterator it = vars.begin(); it != vars.end(); it++)
+   {
+      if (r::sexp::isExternalPointer(it->second) && 
+          r::sexp::isNullExternalPointer(it->second) == nullPtr)
+      {
+         return true;
+      }
+
+      if (r::sexp::isEnvironment(it->second) && level > 0)
+      {
+         // if this object is itself an environment, check it recursively for external pointers. 
+         // (we do this only if there's sufficient recursion depth remaining)
+         if (hasExternalPtr(it->second, nullPtr, level - 1))
+            return true;
+      }
+   }
+
+   return false;
+}
+
+SEXP rs_hasExternalPointer(SEXP objSEXP, SEXP nullSEXP)
+{
+   bool nullPtr = r::sexp::asLogical(nullSEXP);
+   r::sexp::Protect protect;
+   bool hasPtr = false;
+   if (r::sexp::isExternalPointer(objSEXP))
```

Comment:
Should we be checking S4 objects here as well? (the previous implementation recursed into S4 objects but not environments; it seems like we should really be doing both)